code/modules/paperwork/filingcabinet.dm

```diff
@@ -91,23 +83,71 @@
 			I.loc = loc
 			if(prob(25))
 				step_rand(I)
+			SStgui.update_uis(src)
 			to_chat(user, "<span class='notice'>You pull \a [I] out of [src] at random.</span>")
 			return
 	to_chat(user, "<span class='notice'>You find nothing in [src].</span>")
 
-/obj/structure/filingcabinet/Topic(href, href_list)
-	if(href_list["retrieve"])
-		usr << browse(null, "window=filingcabinet") // Close the menu
+/obj/structure/filingcabinet/ui_interact(mob/user, ui_key, datum/tgui/ui, force_open, datum/tgui/master_ui, datum/ui_state/state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "FilingCabinet",  name, 400, 300, master_ui, state)
+		ui.open()
+
+/obj/structure/filingcabinet/ui_data(mob/user)
+	. = ..()
+	var/list/data = list()
+	data["contents"] = null
+	var/list/items = list()
+
+	var/index = 1
+	for(var/obj/item/P in contents)
+		items.Add(list(list("display_name" = capitalize(P.name), "index" = index)))
+		index++
+
+	if(length(items))
+		data["contents"] = items
 
-		//var/retrieveindex = text2num(href_list["retrieve"])
-		var/obj/item/P = locate(href_list["retrieve"])//contents[retrieveindex]
-		if(istype(P) && (P.loc == src) && src.Adjacent(usr))
-			usr.put_in_hands(P)
-			updateUsrDialog()
-			icon_state = "[initial(icon_state)]-open"
-			sleep(5)
-			icon_state = initial(icon_state)
+	return data
+
+/obj/structure/filingcabinet/ui_act(action, params, datum/tgui/ui)
+	if(..())
+		return
+
+	. = TRUE
+
+	add_fingerprint(usr)
+
+	switch(action)
+		if("retrieve")
+			var/index = text2num(params["index"])
+			if(isnull(index) || !ISINDEXSAFE(contents, index))
+				return FALSE
+			var/K = contents[index]
+			retrieve(K)
+
+/obj/structure/filingcabinet/proc/insert(obj/item/O)
+	if(accept_check(O))
+		to_chat(usr, "<span class='notice'>You put [O.name] in [src].</span>")
+		usr.drop_item()
+		O.loc = src
+		icon_state = "[initial(icon_state)]-open"
+		sleep(5)
+		icon_state = initial(icon_state)
+		SStgui.update_uis(src)
+		return TRUE
+
+/obj/structure/filingcabinet/proc/retrieve(obj/item/O)
+	if(istype(O) && (O.loc == src) && src.Adjacent(usr))
+		if(!usr.put_in_hands(O))
+			O.forceMove(loc)
+		SStgui.update_uis(src)
```

Comment:
remove only seems to be called from `ui_act` so we don't need to update it here.
```suggestion
```